File: src/xFrameOptions.ts
/**
 * xFrameOptions.ts — clickjacking defence detector. T76.
 *
 * The page's framing policy decides whether other origins can
 * embed it in an `<iframe>`. Without a policy, an attacker can
 * iframe the site, overlay invisible UI on top, and trick the
 * user into clicking through (clickjacking). Two headers control
 * this:
 *
 *   - `X-Frame-Options` (legacy, but still honoured by every
 *     browser): DENY | SAMEORIGIN | ALLOW-FROM <uri>.
 *   - `Content-Security-Policy: frame-ancestors ...` (modern,
 *     supersedes XFO when present).
 *
 * Either one provides the protection. The detector flags pages
 * that have NEITHER.
 *
 * Findings:
 *
 *   - frame-options.missing             strict
 *     Page has no X-Frame-Options header AND no
 *     `frame-ancestors` directive in its Content-Security-Policy.
 *     Iframable by any origin → clickjacking risk.
 *
 *   - frame-options.allowall            warn
 *     X-Frame-Options is set but `ALLOW-FROM *` (or
 *     equivalent open `frame-ancestors *`). Effectively no
 *     protection — same as missing, but signals an intentional
 *     open policy that may have been a mistake.
 *
 *   - frame-options.invalid             warn
 *     X-Frame-Options is set to a value other than DENY /
 *     SAMEORIGIN / ALLOW-FROM <uri>. Browsers treat unrecognised
 *     values as no-protection.
 *
 * Out of scope:
 *   * http pages — clickjacking still applies but the project
 *     considers http-only sites already broken on transport
 *     security. Don't double-report. Same exemption as hsts.
 *   * Localhost — for dev-only.
 *   * Pages that legitimately want to be embedded everywhere
 *     (oEmbed widgets, embeddable players) should set
 *     `Content-Security-Policy: frame-ancestors *` to make the
 *     intent explicit; that gets the warn (not strict).
 *
 * Reads from the same `topLevelResponseHeaders` Map as
 * hstsHeader — single capture path, second consumer.
 *
 * Mirror: crates/crawler-detectors/src/x_frame_options.rs.
 */

export interface XFrameOptionsFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface XFrameOptionsSnapshot {
  pageUrl: string;
  /** True iff the page itself was loaded over https. */
  pageIsHttps: boolean;
  /** Localhost / loopback exemption. */
  pageIsLocalhost: boolean;
  /** Raw value of the X-Frame-Options header, '' if absent. */
  xfoValue: string;
  /** Raw value of the Content-Security-Policy header, '' if absent. */
  cspValue: string;
}

function isLocalhost(url: string): boolean {
  try {
    const u = new URL(url);
    const h = u.hostname;
    return h === 'localhost' || h === '127.0.0.1' || h === '::1' || h.endsWith('.localhost');
  } catch {
    return false;
  }
}

export function buildXFrameOptionsSnapshot(
  pageUrl: string,
  headers: Record<string, string> | undefined,
): XFrameOptionsSnapshot {
  const pageIsHttps = pageUrl.startsWith('https://');
  const pageIsLocalhost = isLocalhost(pageUrl);
  let xfoValue = '';
  let cspValue = '';
  if (headers) {
    for (const [k, v] of Object.entries(headers)) {
      const lk = k.toLowerCase();
      if (lk === 'x-frame-options') xfoValue = v;
      else if (lk === 'content-security-policy') cspValue = v;
    }
  }
  return { pageUrl, pageIsHttps, pageIsLocalhost, xfoValue, cspValue };
}

/**
 * Extract the `frame-ancestors` directive value from a CSP
 * header. Returns null when absent. CSP directives are
 * semicolon-separated; values within a directive are
 * whitespace-separated source-list tokens.
 */
function parseFrameAncestors(csp: string): string | null {
  for (const part of csp.split(';')) {
    const t = part.trim();
    const m = /^frame-ancestors\s+(.+)$/i.exec(t);
    if (m) return m[1].trim();
  }
  return null;
}

/**
 * Validate an X-Frame-Options value against the spec
 * (RFC 7034). Returns 'deny' | 'sameorigin' | 'allow-from' |
 * 'invalid'. Case-insensitive.
 */
function classifyXfo(value: string): 'deny' | 'sameorigin' | 'allow-from' | 'invalid' {
  const v = value.trim().toLowerCase();
  if (v === 'deny') return 'deny';
  if (v === 'sameorigin') return 'sameorigin';
  if (/^allow-from\s+\S+/.test(v)) return 'allow-from';
  return 'invalid';
}

export function detectXFrameOptionsIssues(
  snap: XFrameOptionsSnapshot,
): XFrameOptionsFinding[] {
  if (!snap.pageIsHttps) return [];
  if (snap.pageIsLocalhost) return [];

  const out: XFrameOptionsFinding[] = [];
  const xfo = snap.xfoValue.trim();
  const csp = snap.cspValue.trim();
  const frameAncestors = csp ? parseFrameAncestors(csp) : null;

  // CSP frame-ancestors supersedes XFO. If present and not
  // open-wildcard, we're protected — no findings.
  if (frameAncestors && frameAncestors !== '*') {
    return [];
  }

  // Flag explicit "allow all" (CSP frame-ancestors * OR XFO
  // allow-from *). The wildcard form is intentional but worth
  // surfacing — operator should confirm it's deliberate.
  if (frameAncestors === '*') {
    out.push({
      severity: 'warn',
      kind: 'frame-options.allowall',
      detail: `Content-Security-Policy frame-ancestors '*' — page is iframable by any origin. If this is an embeddable widget, fine; if it's the main app, it's a clickjacking risk. Consider 'frame-ancestors 'none'' or a specific origin allowlist.`,
      evidence: { pageUrl: snap.pageUrl, cspValue: csp },
    });
    return out;
  }

  // Now check XFO. CSP didn't help (no frame-ancestors at all).
  if (xfo === '') {
    out.push({
      severity: 'strict',
      kind: 'frame-options.missing',
      detail: `Page has no X-Frame-Options header AND no Content-Security-Policy frame-ancestors directive. Any origin can iframe this page → clickjacking attacks possible. Add 'X-Frame-Options: SAMEORIGIN' (legacy + universal) OR 'Content-Security-Policy: frame-ancestors 'self'' (modern).`,
      evidence: { pageUrl: snap.pageUrl },
    });
    return out;
  }

  const cls = classifyXfo(xfo);
  if (cls === 'invalid') {
    out.push({
      severity: 'warn',
      kind: 'frame-options.invalid',
      detail: `X-Frame-Options has an unrecognised value '${xfo}'. Browsers treat invalid values as no-protection. Use DENY (no framing at all) or SAMEORIGIN (only your own origin can iframe).`,
      evidence: { pageUrl: snap.pageUrl, xfoValue: xfo },
    });
  }

  return out;
}

File: crates/crawler-detectors/src/x_frame_options.rs
//! `x_frame_options` — clickjacking-defence detector.
//!
//! Mirror of `src/xFrameOptions.ts`. Findings:
//!
//!   * `frame-options.missing`   strict
//!   * `frame-options.allowall`  warn
//!   * `frame-options.invalid`   warn
//!
//! Either `X-Frame-Options` (RFC 7034) or
//! `Content-Security-Policy: frame-ancestors …` provides
//! clickjacking protection. The detector flags pages with
//! NEITHER, with an open wildcard, or with an unrecognised XFO
//! value.
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector,
//! no I/O, no `unwrap`/`expect` in non-test code.

use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};

/// Captured page state the detector consumes.
///
/// BUG ASSUMPTION: caller is responsible for passing the
/// top-level navigation response headers, NOT subresource
/// headers. XFO/CSP only matter on the document response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct XFrameOptionsSnapshot {
    /// URL of the page (for evidence + http/localhost gates).
    pub page_url: String,
    /// True iff the page itself was loaded over https.
    pub page_is_https: bool,
    /// Localhost / loopback exemption.
    pub page_is_localhost: bool,
    /// Raw `X-Frame-Options` header value, empty string if absent.
    pub xfo_value: String,
    /// Raw `Content-Security-Policy` header value, empty if absent.
    pub csp_value: String,
}

use crate::url_helpers::is_localhost;

/// Build a snapshot from a page URL + a header map. Header keys
/// are looked up case-insensitively (per RFC 9110).
pub fn build_x_frame_options_snapshot<I, K, V>(
    page_url: &str,
    headers: I,
) -> XFrameOptionsSnapshot
where
    I: IntoIterator<Item = (K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{
    let mut xfo = String::new();
    let mut csp = String::new();
    for (k, v) in headers {
        let lk = k.as_ref().to_ascii_lowercase();
        match lk.as_str() {
            "x-frame-options" => xfo = v.as_ref().to_owned(),
            "content-security-policy" => csp = v.as_ref().to_owned(),
            _ => {}
        }
    }
    XFrameOptionsSnapshot {
        page_url: page_url.to_owned(),
        page_is_https: page_url.starts_with("https://"),
        page_is_localhost: is_localhost(page_url),
        xfo_value: xfo,
        csp_value: csp,
    }
}

/// Pull the `frame-ancestors` directive value from a CSP header.
/// Returns `None` when absent. CSP directives are
/// semicolon-separated; values within a directive are
/// whitespace-separated source-list tokens.
fn parse_frame_ancestors(csp: &str) -> Option<String> {
    for part in csp.split(';') {
        let t = part.trim();
        // Match "frame-ancestors <whitespace> <value>" case-insensitively.
        let lower = t.to_ascii_lowercase();
        if let Some(rest) = lower.strip_prefix("frame-ancestors") {
            let after = rest.trim_start();
            if after.is_empty() || (rest.len() == after.len()) {
                continue; // need at least one whitespace separator
            }
            // Use the original casing for the returned value (it
            // may include URL hosts whose case is preserved).
            let original_after = &t[t.len() - after.len()..];
            return Some(original_after.trim().to_owned());
        }
    }
    None
}

#[derive(Debug, PartialEq, Eq)]
enum XfoClass {
    Deny,
    SameOrigin,
    AllowFrom,
    Invalid,
}

/// Validate an X-Frame-Options value against RFC 7034.
fn classify_xfo(value: &str) -> XfoClass {
    let v = value.trim().to_ascii_lowercase();
    if v == "deny" {
        XfoClass::Deny
    } else if v == "sameorigin" {
        XfoClass::SameOrigin
    } else if v.starts_with("allow-from") {
        // Must have whitespace + at least one non-ws char after.
        let after = v.trim_start_matches("allow-from");
        if after.starts_with(|c: char| c.is_ascii_whitespace())
            && after.split_whitespace().next().is_some()
        {
            XfoClass::AllowFrom
        } else {
            XfoClass::Invalid
        }
    } else {
        XfoClass::Invalid
    }
}

/// Run the detector on a snapshot. Returns zero findings when
/// the page is exempt (http or localhost).
pub fn detect_x_frame_options_issues(snap: &XFrameOptionsSnapshot) -> Vec<AxisFinding> {
    if !snap.page_is_https {
        return Vec::new();
    }
    if snap.page_is_localhost {
        return Vec::new();
    }

    let mut out = Vec::new();
    let xfo = snap.xfo_value.trim();
    let csp = snap.csp_value.trim();
    let frame_ancestors = if csp.is_empty() {
        None
    } else {
        parse_frame_ancestors(csp)
    };

    // CSP frame-ancestors supersedes XFO. If present and not the
    // open wildcard, we're protected.
    if let Some(fa) = frame_ancestors.as_deref() {
        if fa != "*" {
            return out;
        }
        // Wildcard: warn, then exit.
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "frame-options.allowall".into(),
            detail: format!(
                "Content-Security-Policy frame-ancestors '*' — page is iframable by any origin. \
                 If this is an embeddable widget, fine; if it's the main app, it's a clickjacking risk. \
                 Consider 'frame-ancestors 'none'' or a specific origin allowlist. (cspValue={csp})"
            ),
        });
        return out;
    }

    if xfo.is_empty() {
        out.push(AxisFinding {
            severity: AxisSeverity::Strict,
            kind: "frame-options.missing".into(),
            detail: format!(
                "Page has no X-Frame-Options header AND no Content-Security-Policy frame-ancestors directive. \
                 Any origin can iframe this page → clickjacking attacks possible. \
                 Add 'X-Frame-Options: SAMEORIGIN' OR 'Content-Security-Policy: frame-ancestors 'self''. \
                 (pageUrl={url})",
                url = snap.page_url
            ),
        });
        return out;
    }

    if classify_xfo(xfo) == XfoClass::Invalid {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "frame-options.invalid".into(),
            detail: format!(
                "X-Frame-Options has an unrecognised value '{xfo}'. \
                 Browsers treat invalid values as no-protection. \
                 Use DENY (no framing at all) or SAMEORIGIN (only your own origin can iframe)."
            ),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap(url: &str, xfo: &str, csp: &str) -> XFrameOptionsSnapshot {
        build_x_frame_options_snapshot(
            url,
            [("x-frame-options", xfo), ("content-security-policy", csp)],
        )
    }

    #[test]
    fn http_page_is_exempt() {
        let s = snap("http://example.com/", "", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn localhost_is_exempt() {
        let s = snap("http://localhost:8080/", "", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
        let s = snap("https://localhost/", "", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
        let s = snap("https://127.0.0.1/", "", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn dot_localhost_subdomain_is_exempt() {
        let s = snap("https://app.localhost/", "", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn missing_both_strict() {
        let s = snap("https://example.com/", "", "");
        let f = detect_x_frame_options_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "frame-options.missing");
        assert_eq!(f[0].severity, AxisSeverity::Strict);
    }

    #[test]
    fn xfo_deny_alone_is_protection() {
        let s = snap("https://example.com/", "DENY", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn xfo_sameorigin_alone_is_protection() {
        let s = snap("https://example.com/", "SAMEORIGIN", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn xfo_lowercase_normalized() {
        let s = snap("https://example.com/", "sameorigin", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn xfo_invalid_value_warn() {
        let s = snap("https://example.com/", "BANANA", "");
        let f = detect_x_frame_options_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "frame-options.invalid");
        assert_eq!(f[0].severity, AxisSeverity::Warn);
    }

    #[test]
    fn xfo_allow_from_with_uri_is_valid() {
        let s = snap("https://example.com/", "ALLOW-FROM https://parent.example", "");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn xfo_allow_from_no_uri_is_invalid() {
        let s = snap("https://example.com/", "ALLOW-FROM", "");
        let f = detect_x_frame_options_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "frame-options.invalid");
    }

    #[test]
    fn csp_frame_ancestors_self_supersedes_missing_xfo() {
        let s = snap("https://example.com/", "", "frame-ancestors 'self'");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn csp_frame_ancestors_none_supersedes_missing_xfo() {
        let s = snap("https://example.com/", "", "frame-ancestors 'none'");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn csp_frame_ancestors_wildcard_warn() {
        let s = snap("https://example.com/", "", "frame-ancestors *");
        let f = detect_x_frame_options_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "frame-options.allowall");
        assert_eq!(f[0].severity, AxisSeverity::Warn);
    }

    #[test]
    fn csp_frame_ancestors_in_middle_directive() {
        let s = snap(
            "https://example.com/",
            "",
            "default-src 'self'; frame-ancestors 'self'; script-src 'self'",
        );
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn csp_with_other_directives_no_frame_ancestors_uses_xfo() {
        // No frame-ancestors → fall through to XFO check (which is
        // missing) → strict.
        let s = snap("https://example.com/", "", "default-src 'self'");
        let f = detect_x_frame_options_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "frame-options.missing");
    }

    #[test]
    fn csp_frame_ancestors_case_insensitive_directive_name() {
        let s = snap("https://example.com/", "", "FRAME-ANCESTORS 'self'");
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn header_lookup_is_case_insensitive() {
        let s = build_x_frame_options_snapshot(
            "https://example.com/",
            [("X-Frame-Options", "DENY")],
        );
        assert_eq!(detect_x_frame_options_issues(&s).len(), 0);
    }

    #[test]
    fn wildcard_csp_returns_only_warn_not_strict() {
        let s = snap("https://example.com/", "", "frame-ancestors *");
        let f = detect_x_frame_options_issues(&s);
        assert!(!f.iter().any(|x| x.kind == "frame-options.missing"));
    }

    #[test]
    fn detail_text_includes_evidence_url() {
        let s = snap("https://acme.example.com/foo", "", "");
        let f = detect_x_frame_options_issues(&s);
        assert!(f[0].detail.contains("acme.example.com"));
    }

    #[test]
    fn malformed_url_is_treated_as_not_localhost() {
        // url::Url::parse will fail on "https://" alone; that's
        // fine — the host check returns false → not localhost.
        // Page-https check still triggers via str::starts_with.
        let s = snap("https://", "", "");
        let f = detect_x_frame_options_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "frame-options.missing");
    }
}
